File: mobile/src/types/resource.ts
import type { User } from "./auth";

export const RESOURCE_CATEGORIES = [
  "NOTES",
  "ASSIGNMENTS",
  "PREVIOUS_PAPERS",
  "PPTS",
  "LAB_RECORDS",
  "QUESTION_BANKS",
  "OTHER",
] as const;

export type ResourceCategory = (typeof RESOURCE_CATEGORIES)[number];
export type ResourceVisibility = "COMMUNITY" | "PUBLIC";
export type ResourceSort = "recent" | "downloads" | "name";

export interface Resource {
  _id: string;
  title: string;
  description: string;
  fileName: string;
  fileUrl: string;
  fileSize: number;
  fileType: string;
  category: ResourceCategory;
  tags: string[];
  uploadedBy: Pick<User, "_id" | "fullName" | "rollNumber" | "profilePicture">;
  communityId: { _id: string; name: string; slug: string };
  downloadCount: number;
  visibility: ResourceVisibility;
  createdAt: string;
  updatedAt: string;
}

export interface PaginatedResources {
  items: Resource[];
  total: number;
  page: number;
  limit: number;
  pages: number;
}

export interface ResourceFormData {
  title: string;
  description: string;
  category: ResourceCategory;
  tags: string[];
  visibility: ResourceVisibility;
}

File: backend-java/src/main/java/com/studyconnect/backend/entity/Resource.java
package com.studyconnect.backend.entity;

import com.studyconnect.backend.entity.enums.ResourceCategory;
import com.studyconnect.backend.entity.enums.ResourceVisibility;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.CompoundIndex;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.FieldType;
import org.springframework.data.mongodb.core.mapping.MongoId;

@Getter
@Setter
@NoArgsConstructor
@Document(collection = "resources")
@CompoundIndex(name = "resource_category_community_created", def = "{'category': 1, 'communityId': 1, 'createdAt': -1}")
public class Resource {

    @Id
    @MongoId(FieldType.OBJECT_ID)
    private String id;

    private String title;
    private String description = "";
    private String fileName;
    private String fileUrl;
    private long fileSize;
    private String fileType;

    @Indexed
    private ResourceCategory category;
    private List<String> tags = new ArrayList<>();

    @Indexed
    private String uploadedBy;

    @Indexed
    private String communityId;

    private long downloadCount = 0;
    private ResourceVisibility visibility = ResourceVisibility.COMMUNITY;

    @CreatedDate
    private Instant createdAt;

    @LastModifiedDate
    private Instant updatedAt;
}
